client/v2/tx/tx.go

```diff
@@ -0,0 +1,265 @@
+package tx
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cosmos/cosmos-sdk/client/input"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	gogogrpc "github.com/cosmos/gogoproto/grpc"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/spf13/pflag"
+	"os"
+)
+
+// GasEstimateResponse defines a response definition for tx gas estimation.
+type GasEstimateResponse struct {
+	GasEstimate uint64 `json:"gas_estimate" yaml:"gas_estimate"`
+}
+
+func (gr GasEstimateResponse) String() string {
+	return fmt.Sprintf("gas estimate: %d", gr.GasEstimate)
+}
+
+// GenerateOrBroadcastTxCLI will either generate and print an unsigned transaction
+// or sign it and broadcast it returning an error upon failure.
+func GenerateOrBroadcastTxCLI(clientCtx txContext, flagSet *pflag.FlagSet, msgs ...sdk.Msg) error {
+	txf, err := NewFactoryCLI(clientCtx, flagSet)
+	if err != nil {
+		return err
+	}
+
+	return GenerateOrBroadcastTxWithFactory(clientCtx, txf, msgs...)
+}
+
+// GenerateOrBroadcastTxWithFactory will either generate and print an unsigned transaction
+// or sign it and broadcast it returning an error upon failure.
+func GenerateOrBroadcastTxWithFactory(clientCtx txContext, txf Factory, msgs ...sdk.MsgV2) error {
+	// Validate all msgs before generating or broadcasting the tx.
+	// We were calling ValidateBasic separately in each CLI handler before.
+	// Right now, we're factorizing that call inside this function.
+	// ref: https://github.com/cosmos/cosmos-sdk/pull/9236#discussion_r623803504
+	for _, msg := range msgs {
+		m, ok := msg.(sdk.HasValidateBasic)
+		if !ok {
+			continue
+		}
+
+		if err := m.ValidateBasic(); err != nil {
+			return err
+		}
+	}
+
+	// If the --aux flag is set, we simply generate and print the AuxSignerData.
+	if clientCtx.IsAux {
+		auxSignerData, err := makeAuxSignerData(clientCtx, txf, msgs...)
+		if err != nil {
+			return err
+		}
+
+		return clientCtx.PrintProto(&auxSignerData)
+	}
+
+	if clientCtx.GenerateOnly {
+		return txf.PrintUnsignedTx(clientCtx, msgs...)
+	}
+
+	return BroadcastTx(clientCtx, txf, msgs...)
+}
+
+// BroadcastTx attempts to generate, sign and broadcast a transaction with the
+// given set of messages. It will also simulate gas requirements if necessary.
+// It will return an error upon failure.
+func BroadcastTx(clientCtx txContext, txf Factory, msgs ...sdk.MsgV2) error {
```

Comment:
Also this func is very large to be only for broadcasting, could be refactored to only broadcast the final Tx. Most logic here could me moved inside `Factory`